### PR TITLE
coins python challenge

### PR DESCRIPTION
**Pull request by abed42**
- [x] PR's title includes the challenge's name and language
- [x] PR has short and clear description of the challenge
- [x] PR has appropriate labels and milestones for easy identification
- [x] PR it is assigned to the owner
- [x] reviewers are assigned
- [x] the PR contributes only one focused change
- [x] the branch is up to date with `main`/`master`
- [x] the code works when pulled and run locally
- [x] all conflicts are resolved (if any)
- [x] It is linked to an issue in the appropriate column of the project board
- [x] feedback is addressed (if any, and if it is appropriate feedback.)

## README Documentation

- [x] The solution is documented in `/src/README.md`
- [x] The markdown source is formatted
- [x] Spelling and grammar is correct in all text
- [x] The markdown looks correct when you preview the file
- [x] All links and images work
- [x] The README documents the solution's behavior, strategy(ies) and implementation(s)
- [x] There are use cases in code blocks to illustrate the function's behavior

## Python Files

- [x] There is a module header
- [x] There is a module docstring
- [x] File names are in snake_case
- [x] Test files are named `test_<module_name>.py`

## Function Docstring

- [x] Behavior description
- [x] Parameter description
- [x] Return value description
- [x] Include any assumptions
- [x] Include 3 or more (passing!) doctests
- [x] Include 1-2 use cases (if necessary)

## Function Implementation

- [x] The solution is not is not copied from [djeada](https://github.com/djeada/Algorithms-And-Data-Structures)
- [x] The code is auto-formatted
- [x] The code has no (reasonable) linting mistakes
- [x] Variables are named with snake_case
- [x] The function has a clear and helpful name
- [x] The file's name matches the function name
- [x] The code follows the strategy as simply as possible
- [x] Variable names are clear and helpful
- [x] Comments explain the strategy (if necessary)
- [x] There are type annotations
- [x] (challenge) The code includes defensive assertions

## Unit Test Suite

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Each unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases
- [x] Test suite includes black-box and glass-box tests

### CODE REVIEW

File: src/dynamic_programming/python/coins/src/coin_change.py [hunk 1]
```diff
@@ -1,28 +1,87 @@
...
+    Parameters:
+        amount (int): The target amount to be made up using the coins.
+        coins (list): A list of coin denominations.
+
+    Returns:
+        int: The minimum number of coins needed to make up the target amount. Returns -1 if it is not possible to make up the target amount using the given coins.
+    """
+    memo = {}
+
+    def dp(n):
```
Comment (mohammedalasly) [inline src/dynamic_programming/python/coins/src/coin_change.py]: Please provide a more descriptive function name(make sure it's in snake_case)

**Review by mohammedalasly — CHANGES_REQUESTED** [review]

File: src/dynamic_programming/python/coins/src/coin_change.py (hunk 1 above)
Comment (abed42) [reply src/dynamic_programming/python/coins/src/coin_change.py]: yea good catch! I'll update that one!
